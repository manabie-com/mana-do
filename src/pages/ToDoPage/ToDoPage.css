h1.title {
	font-size: 120px;
    text-align: center;
    color: darkseagreen;
    padding: 2rem;
}

.ToDo__container {
	display: flex;
	width: 100% !important;
	justify-content: center;
	flex-direction: column;
    background: white;
	box-shadow: 0 2px 4px 0 rgb(0 0 0 / 20%), 0 25px 50px 0 rgb(0 0 0 / 10%);
    flex: 0 0 50%;
    max-width: 50%;
}

.ToDo__list {
	display: flex;
	flex-direction: column;
}

.ToDo__empty-msg {
	min-height: 40px;
	display: flex;
	align-items: center;
	padding: 0 30px;
	font-style: italic;
}

@media (max-width: 1500px) {
	.ToDo__container {
		flex: 0 0 60%;
		max-width: 60%;
	}
}

@media (max-width: 1100px) {
	.ToDo__container {
		flex: 0 0 70%;
		max-width: 70%;
	}
}

@media (max-width: 900px) {
	.ToDo__container {
		flex: 0 0 95%;
		max-width: 95%;
	}
}