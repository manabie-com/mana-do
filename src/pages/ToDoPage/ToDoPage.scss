@import "../../assets/scss/variables";

.to-do {
  display: flex;
  flex-direction: column;
  align-items: center;
  &__main {
    width: 100%;
    max-width: 600px;
    margin-bottom: 24px;
  }
  &__container {
    padding: 0 16px;
    width: 100%;
    display: flex;
    flex-direction: column;
    align-items: center;
  }
}
