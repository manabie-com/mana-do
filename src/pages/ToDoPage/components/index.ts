export * from "./ItemTodo/ItemTodo";
export * from "./LabelTodo/LabelTodo";
export * from "./Toolbar/Toolbar";
export * from "./Banner/Banner";
