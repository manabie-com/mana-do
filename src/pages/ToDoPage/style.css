.ToDo__container {
  display: flex;
  justify-content: center;
  width: 100%;
  min-height: 100vh;
  background: var(--primary-color-lightest);
}
.ToDo_content {
  width: 50%;
}

.ToDo_content__header {
  color: var(--primary-color);
  font-weight: bold;
  font-size: 30px;
}
