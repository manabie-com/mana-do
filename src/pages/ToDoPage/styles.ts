import styled from "styled-components";

export const Wrapper = styled.div`
  width: 100%;

  .title {
    color: var(--font-color);
    text-align: center;
    font-size: 40px;
    font-weight: 700;
    margin-top: 40px;
  }

  .list-container {
    margin: 0 auto;
    width: 540px;
    padding: 50px 0;

    .input-wrap {
      background: var(--clr-white);
      border-radius: var(--radius);
      box-shadow: var(--light-shadow);
      transition: var(--transition);
      margin: auto;

      .form-control {
        display: flex;
        justify-content: center;
        height: 65px;
        margin-bottom: 25px;

        input {
          padding: 24px;
          background: var(--listbg);
          border-radius: 5px;
          border-color: transparent;
          flex: 1 0 auto;
          color: var(--font-color);
          font-size: 14px;
          font-family: inherit;

          :focus,
          :active {
            outline: none;
          }
        }
      }
    }
  }

  @media screen and (max-width: 768px) {
    .title {
      font-size: 30px;
      margin-top: 24px;
    }

    .list-container {
      width: 100%;
      padding: 24px 0;
    }
  }
`;
