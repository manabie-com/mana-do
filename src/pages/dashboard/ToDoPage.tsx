import React, { useContext, useEffect, useRef, useState } from 'react';
import { RouteComponentProps } from 'react-router-dom';
import {
    setTodos,
    createTodo,
    deleteAllTodos,
    actSignOut,
} from '../../store/actions';
import Service from '../../service';
import { TodoStatus } from '../../models/todo';
import { AppContext } from '../../store/context';
import AppButton from '../../commons/AppButton';
import AppTitle from '../../components/AppTitle';
import TodoList from '../../components/TodoList';
import AppRow from '../../commons/AppRow';
import AppCol from '../../commons/AppCol';
import AppText from '../../commons/AppText';
import AppInput from '../../commons/AppInput';
import { useNotAuthenticated } from '../../hooks/useNotAuthenticated';
import AppFooter from '../../components/AppFooter';

// type EnhanceTodoStatus = TodoStatus | 'ALL';


const ToDoPage = ({ history }: RouteComponentProps) => {
    useNotAuthenticated();
    const { state, dispatch } = useContext(AppContext);
    const { auth } = state;
    const inputRef = useRef<HTMLInputElement>(null);
    const [activeTab, setActiveTab] = useState(localStorage.getItem('filter')||'ALL');
    useEffect(() => {
        (async () => {
            try {
                const resp = await Service.getTodos();
                dispatch(setTodos(resp.data || []));

            } catch (error) {
                if (error.status === 401) {
                    history.push('/')
                }
            }
        })()
    }, [dispatch, history])

    const onCreateTodo = async (e: React.KeyboardEvent<HTMLInputElement>) => {
        if (e.key === 'Enter' && inputRef.current) {

            try {
                const resp = await Service.createTodo(inputRef.current.value);
                dispatch(createTodo(resp.data));
                inputRef.current.value = '';
            } catch (e) {
                if (e.status === 401) {
                    history.push('/')
                }
            }
        }
    }

    const onDeleteAllTodo = async () => {
        dispatch(deleteAllTodos());
        try {
            const response = await Service.clearTodos();
            console.log(response);
        } catch (error) {
            if (error.status === 401) {
                history.push('/');
            }
        }
    }

    const filterButtonsJSX = ():React.ReactNode => {
        return (<>
        <AppButton btnType={activeTab==='ALL'?'primary':'default'}  key="ALL" onClick={()=>{setActiveTab('ALL'); localStorage.setItem('filter','ALL')}}>All</AppButton>
        {Object.entries(TodoStatus).map(item=><AppButton key={item[0]} btnType={activeTab===item[1]?'primary':'default'} onClick={(e:React.MouseEventHandler<HTMLButtonElement>)=>{setActiveTab(item[1]); localStorage.setItem('filter',item[1])}}>{item[0]}</AppButton>)}
        </>)
    }

    const onClickSignOut = () => {
        dispatch(actSignOut());
        localStorage.removeItem('token');
        history.push('/');
    }

    return (
        <AppRow className="justify-content-center">
            <AppCol xs={5} sm ={9} md={10} lg={8} xl={8}>
            <div className="ToDo__container">
                <AppTitle />
                <p>Hi, <AppText strong>{auth.currentUser?.firstName || ''}</AppText><br /><AppButton btnType="link" href="#" onClick={onClickSignOut}>Sign out</AppButton></p>
                <div className="Todo__creation">
                    <AppInput
                        ref={inputRef}
                        className="Todo__input"
                        placeholder="What need to be done?"
                        onKeyDown={onCreateTodo}
                        style={{marginBottom: 20}}
                    />
                </div>                
                    <TodoList filter={activeTab}></TodoList>
                <div className="Todo__toolbar">
                    <div className="Todo__tabs">
                        {filterButtonsJSX()}
                    </div>
                    <AppButton btnType="primary" onClick={onDeleteAllTodo}>
                        Clear all
                    </AppButton>
                </div>
            </div>
            <AppFooter textColor="white" />
            </AppCol>
        </AppRow>
    );
};

export default ToDoPage;