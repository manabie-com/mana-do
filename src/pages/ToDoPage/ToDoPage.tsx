import React, {useEffect, useReducer, useRef, useState} from 'react';

import reducer, {initialState} from '../../store/reducer';
import {
    setTodos,
    createTodo,
    toggleAllTodos,
    deleteAllTodos,
    updateTodoStatus,
    deleteTodo
} from '../../store/actions';
import Service from '../../service';
import './ToDoPage.css'
import { EnhanceTodoStatus, Todo } from '../../models'
import ToDoItem, { ToDoItemProps } from '../../components/ToDoItem/ToDoItem';
import ToDoInput from '../../components/ToDoInput/ToDoInput';
import ToDoToolbar, { ToDoToolbarProps } from '../../components/ToDoToolbar/ToDoToolbar';



const ToDoPage = () => {
    const [{todos}, dispatch] = useReducer(reducer, initialState);
    const [showing, setShowing] = useState<EnhanceTodoStatus>('ALL');
    const inputRef = useRef<any>(null);

    useEffect(() => {
        (async () => {
            const resp = await Service.getTodos();

            dispatch(setTodos(resp));
        })()
    }, []);

    // Update localStorage every todos has changed
    useEffect(()=>{
        (async ()=>{
            // Make sure todos is available
            if (todos.length >= 0) {
                await Service.saveTodos(todos);
            }
        })()
    }, [todos]);

    const onCreateTodo = async (e: React.KeyboardEvent<HTMLInputElement>) => {
        // Make sure that todo input is not empty or all white space
        if (e.key === 'Enter' && inputRef.current.value.trim()) {
            const resp = await Service.createTodo(inputRef.current.value);
            dispatch(createTodo(resp));
            inputRef.current.value = '';
        }
    }

    // Limit use any type
    const onUpdateTodoStatus = (e: React.ChangeEvent<HTMLInputElement>, todoId: string) => {
        dispatch(updateTodoStatus(todoId, e.target.checked))
    }

    const onToggleAllTodo = (e: React.ChangeEvent<HTMLInputElement>) => {
        dispatch(toggleAllTodos(e.target.checked))
    }

    const onDeleteAllTodo = () => {
        dispatch(deleteAllTodos());
    }

    const onDeleteTodo = (todoId: string) => {
        dispatch(deleteTodo(todoId));
    }

    // define props for components
    const toDoToolbarProps: ToDoToolbarProps = {todos, showing, onToggleAllTodo, setShowing, onDeleteAllTodo};

    return (
        <div className="ToDo__container">
            <ToDoInput inputRef={inputRef} onCreateTodo={onCreateTodo}/>
            <div className="ToDo__list">
                {   
                    // filter buttons work incorrectly cause todos always render all todos
                    // Should filter by showing before render
                    todos.filter((todo: Todo) => showing === 'ALL' ? todo : (showing === todo.status))
                        .map((todo: Todo) => {
                            const ToDoItemProps: ToDoItemProps = {todo, onDeleteTodo, onUpdateTodoStatus};
                            return <ToDoItem key={todo.id} {...ToDoItemProps}/>
                        })
                }
            </div>
            <ToDoToolbar {...toDoToolbarProps} />
        </div>
    );
};

export default ToDoPage;