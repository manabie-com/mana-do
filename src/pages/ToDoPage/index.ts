import ToDoPage from "./ToDoPage";

export default ToDoPage;