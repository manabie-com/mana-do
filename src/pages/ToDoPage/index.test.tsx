import * as React from "react";
import renderer from "react-test-renderer";
import { createRoot } from "react-dom/client";

import * as AppContext from "../../AppContext";
import ToDoPage from ".";
import { TodoStatus } from "../../models/todo";

jest
  .mock("react", () => ({
    ...jest.requireActual("react"),
    useState: jest.fn(),
  }))
  .mock("../../components/Footer", () => "Footer")
  .mock("../../components/todos/TodoList", () => "TodoList")
  .mock("../../layouts/main", () => "Layout");

let container: any = null;
let root: any = null;
const mockDispatch = jest.fn();

beforeEach(() => {
  const mockContextValue = {
    state: {
      todos: [],
    },
    dispatch: mockDispatch,
  };
  jest
    .spyOn(AppContext, "useAppContext")
    // @ts-ignore
    .mockImplementation(() => mockContextValue);

  container = document.createElement("div");
  root = createRoot(container);
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  root.unmount();
  container.remove();
  container = null;
});

describe("snapshot", () => {
  test("should return snapshot", () => {
    const tree = renderer.create(<ToDoPage />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("create todo list", () => {
  const setTodosValue = jest.fn();
  const mockedTodos = [
    {
      id: "hfuwyr",
      content: "todo 1",
      user_id: "firstUser",
      status: TodoStatus.Active,
      created_date: "",
    },
    {
      id: "orislaj",
      content: "todo 2",
      user_id: "firstUser",
      status: TodoStatus.Active,
      created_date: "",
    },
  ];
  const useStateMock = () => [mockedTodos, setTodosValue];

  // @ts-ignore
  jest.spyOn(React, "useState").mockImplementation(useStateMock);
});
