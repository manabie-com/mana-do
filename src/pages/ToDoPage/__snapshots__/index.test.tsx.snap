// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot should return snapshot 1`] = `
<div
  className="sc-bczRLJ bWAoni"
>
  <div
    className="title"
  >
    Draggable Todos
  </div>
  <div
    className="list-container"
  >
    <div
      className="input-wrap"
    >
      <div
        className="form-control"
      >
        <input
          onKeyDown={[Function]}
          placeholder="Add a Task"
          type="text"
        />
      </div>
    </div>
    <TodoList />
    <Footer />
  </div>
</div>
`;
